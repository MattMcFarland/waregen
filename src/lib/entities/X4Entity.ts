import { Parser, Builder, BuilderOptions } from "../utils/xml";
import { cloneDeep, merge, dig } from "../utils/object";
import idx from "idx";
import { IndexByString } from "@@/utils/object/types";

export enum X4EntityType {
  BASE_ENTITY = "BASE_ENTITY" /* x */,
  HASH = "HASH" /* x */,
  LIBRARY_BASKET = "LIBRARY_BASKET" /* x */,
  LIBRARY_BASKETS = "LIBRARY_BASKETS" /* x */,
  LIBRARY_ICON = "LIBRARY_ICON" /* x */,
  LIBRARY_ICONS = "LIBRARY_ICONS" /* x */,
  LIBRARY_MODULE = "LIBRARY_MODULE" /* x */,
  LIBRARY_MODULES = "LIBRARY_MODULES" /* x */,
  LIBRARY_MODULE_GROUP = "LIBRARY_MODULE_GROUP" /* x */,
  LIBRARY_MODULE_GROUPS = "LIBRARY_MODULE_GROUPS" /* x */,
  LIBRARY_WARE = "LIBRARY_WARE" /* x */,
  LIBRARY_WARES = "LIBRARY_WARES" /* x */,
  ASSET_WARE_MACRO = "ASSET_WARE_MACRO" /* x */,
  ASSET_WARE_PROD_MACRO = "ASSET_WARE_PROD_MACRO" /* x */
}
export interface XMLAttributes {
  Attributes: IndexByString;
}
export type XMLObjectLike = IndexByString[] & XMLAttributes;

export class X4Entity<T> {
  private entityType: X4EntityType = X4EntityType.BASE_ENTITY;
  private create: (options: any) => any;

  protected xmlDef: T = <any>{};
  protected onImport(xmlDef: any): void {}
  protected buildOptions: BuilderOptions;
  protected xmlBuild(
    obj: any,
    pretty: boolean = true,
    options?: BuilderOptions
  ): string {
    const buildOptions = Object.assign({}, this.buildOptions, options);
    if (pretty) {
      buildOptions.renderOpts = {
        pretty: true,
        indent: "  ",
        newline: "\n"
      };
    } else {
      buildOptions.renderOpts = {
        pretty: false,
        indent: undefined,
        newline: undefined
      };
    }
    return new Builder(buildOptions).buildObject(obj);
  }

  public rootName: string;
  public get selector(): string | undefined {
    const keys = Object.keys(this.xmlDef);

    if (keys.length < 1) return undefined;

    const id = this.id;

    if (id) {
      return `*/${this.rootName}[@id='${id}']`;
    } else {
      return undefined;
    }
  }

  public get id() {
    return <string>idx(this.__xmlDef, _ => _.Attributes.id);
  }
  public get __entityType() {
    return this.entityType;
  }
  public get __xmlDef() {
    return JSON.parse(this.toJson());
  }
  public async import(xmlString: string, rootNode?: string): Promise<this> {
    if (!rootNode) {
      this.xmlDef = await (<T>(<any>new Parser({
        explicitRoot: false
      }).parseString(xmlString)));
      typeof this.onImport === "function" && this.onImport(this.__xmlDef);
      return this;
    }
    const _fullObj = await new Parser({
      explicitRoot: true,
      rootName: rootNode
    }).parseString(xmlString);
    this.xmlDef = dig(_fullObj, rootNode)[0];
    this.buildOptions.rootName = rootNode;
    typeof this.onImport === "function" && this.onImport(this.__xmlDef);
    return this;
  }
  public clone(): X4Entity<T> {
    return cloneDeep(this);
  }
  public merge(obj: X4Entity<T>): X4Entity<T> {
    if (obj.__entityType !== this.__entityType) {
      throw new TypeError(
        `cannot merge entityTypes ${obj.__entityType} into ${this.__entityType}`
      );
    }
    return merge(cloneDeep(this), obj);
  }
  public toXml(pretty: boolean = true, options?: BuilderOptions): string {
    return this.xmlBuild(this.xmlDef, pretty, options);
  }
  public toXmlPatch(
    patchType: XMLPatchTypes,
    selector?: string,
    pretty: boolean = true,
    options: BuilderOptions = {}
  ) {
    const obj = {
      [patchType]: [
        {
          Attributes: {
            sel: selector || this.selector
          },
          ...this.__xmlDef
        }
      ]
    };
    return this.xmlBuild(obj, pretty, options);
  }
  // public toXmlPatch(pretty: boolean = true, options: BuilderOptions = {}) {
  //   options.rootName = "diff";
  //   return this.xmlBuild(this.__xmlPatchDef, pretty, options);
  // }

  public toJson(...stringifyOptions: any): string {
    return JSON.stringify(this.xmlDef, ...stringifyOptions);
  }

  constructor(
    entityType: X4EntityType,
    rootName: string,
    defaults: any,
    options?: any
  ) {
    this.buildOptions = {
      headless: true,
      rootName: rootName,
      renderOpts: { pretty: false, indent: undefined, newline: undefined }
    };
    this.create = _options => {
      if (typeof options === "undefined") {
        this.xmlDef = cloneDeep(defaults);
        return this;
      }
      if (options) {
        this.xmlDef = cloneDeep(_options);
      }
      return this;
    };
    this.create(options);
    this.entityType = entityType;
    this.rootName = rootName;
  }
}

export default X4Entity;

export enum XMLPatchTypes {
  ADD = "add",
  REPLACE = "replace",
  REMOVE = "remove"
}
