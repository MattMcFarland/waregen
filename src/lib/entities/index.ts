export { AssetWareMacro } from "./AssetWareMacro";
export { AssetWareProdMacro } from "./AssetWareProdMacro";
export { Basket } from "./Basket";
export { BasketCollection } from "./BasketCollection";
export { Icon } from "./Icon";
export { IconCollection } from "./IconCollection";
export { IndexHash } from "./IndexHash";
export { ModuleGroup } from "./ModuleGroup";
export { ModuleGroupsCollection } from "./ModuleGroupsCollection";
export { Ware } from "./Ware";
export { WareCollection } from "./WareCollection";
export { X4Entity } from "./X4Entity";
