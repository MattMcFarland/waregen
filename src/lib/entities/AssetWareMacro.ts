import { X4Entity, X4EntityType } from "./X4Entity";
import { Macros } from "..//XMLTypes/X4WareMacro";
import idx from "idx";

export class AssetWareMacro extends X4Entity<Macros> {
  get name() {
    return idx(this.xmlDef, _ => _.macro[0].Attributes.name) || "";
  }
  set name(v: string) {
    this.xmlDef = filledXmlObject(v, this.ref);
  }
  get ref() {
    return idx(this.xmlDef, _ => _.macro[0].component[0].Attributes.ref) || "";
  }
  set ref(v: string) {
    this.xmlDef = filledXmlObject(this.name, v);
  }
  constructor(name: string, ref: string) {
    super(X4EntityType.ASSET_WARE_MACRO, "macros", filledXmlObject(name, ref));
  }
}

function filledXmlObject(name: string, ref: string): Macros {
  if (undefined == name) throw new TypeError("name required");
  if (undefined == ref) throw new TypeError("ref required");
  return {
    macro: [
      {
        Attributes: { name, class: "object" },
        component: [{ Attributes: { ref } }],
        properties: [{ identification: [{ Attributes: { unique: 0 } }] }]
      }
    ]
  };
}

export default AssetWareMacro;
