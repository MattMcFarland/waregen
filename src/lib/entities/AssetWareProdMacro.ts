import { X4Entity, X4EntityType } from "./X4Entity";
import { Macros } from "@@/XMLTypes/X4WareMacro";
import idx from "idx";

export class AssetWareProdMacro extends X4Entity<Macros> {
  get name() {
    return idx(this.xmlDef, _ => _.macro[0].Attributes.name) || "";
  }
  set name(v: string) {
    if (this.xmlDef.macro && this.xmlDef.macro[0]) {
      this.xmlDef.macro[0].Attributes.name = v;
    }
  }
  constructor(xmlObj: any = {}) {
    super(X4EntityType.ASSET_WARE_MACRO, "macros", xmlObj);
  }
}

export default AssetWareProdMacro;
