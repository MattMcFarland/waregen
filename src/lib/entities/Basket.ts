import idx from "idx";
import { X4Entity, X4EntityType } from "./X4Entity";
import {
  BasketEntity,
  WareEntity as BasketWareEntity
} from "@@/XMLTypes/X4LibraryBaskets";
import { clone } from "../../lib/utils/object";
export type BasketWareEntity = BasketWareEntity;
export type BasketWareEntities = BasketWareEntity[];

export interface BasketInterface {
  id?: string;
  name?: string;
  description?: string;
  wares?: string[];
}
export class Basket extends X4Entity<BasketEntity> {
  private proxy: any;
  private _wares: string[] = [];
  onImport(xmlDef: BasketEntity) {}
  set id(v: string) {
    this.xmlDef.Attributes.id = v;
  }
  get id() {
    return this.xmlDef.Attributes.id;
  }
  set name(v: string) {
    this.xmlDef.Attributes.name = v;
  }
  get name() {
    return <string>idx(this.xmlDef, _ => _.Attributes.name);
  }
  set description(v: string) {
    this.xmlDef.Attributes.description = v;
  }
  get description() {
    return <string>idx(this.xmlDef, _ => _.Attributes.description);
  }
  get wares() {
    return this.proxy;
  }
  set wares(v: string[]) {
    const xmlDef = this.xmlDef;
    const handler = {
      set(target: string[], key: string, value: string) {
        xmlDef.wares = [
          {
            ware: [
              ...((xmlDef.wares && xmlDef.wares[0] && xmlDef.wares[0].ware) ||
                []),
              mapName(value)
            ]
          }
        ];
        return Reflect.set(target, key, value);
      },
      deleteProperty(target: string[], key: string) {
        return Reflect.deleteProperty(target, key);
      }
    };
    this.proxy = new Proxy(this._wares, handler);
    Object.assign(this.proxy, v);
  }

  constructor(basketInterface?: BasketInterface, xmlDefOptions?: BasketEntity) {
    super(
      X4EntityType.LIBRARY_BASKET,
      "basket",
      xmlDefOptions || getBasketDefaults()
    );
    this.wares = [];
    if (basketInterface) {
      if (basketInterface.description)
        this.description = basketInterface.description;
      if (basketInterface.id) this.id = basketInterface.id;
      if (basketInterface.name) this.name = basketInterface.name;
      if (basketInterface.wares) this.wares = basketInterface.wares;
    }
  }
}

export function getBasketDefaults() {
  return clone(<BasketEntity>(<unknown>{
    Attributes: {}
  }));
}

export function mapName(name: string): any {
  return {
    Attributes: {
      ware: name
    }
  };
}

export function findInXmlDef(xmlDef: BasketEntity, value: string) {
  const waresArray = <BasketWareEntities>idx(xmlDef, _ => _.wares[0].ware);
  return waresArray.find((e: BasketWareEntity) => e.Attributes.ware === value);
}

export default Basket;
