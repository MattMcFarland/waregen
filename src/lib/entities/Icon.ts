import { X4Entity, X4EntityType } from "./X4Entity";
import { clone } from "lodash";

import {
  IconEntity,
  Attributes as IconAttributes
} from "@@/XMLTypes/X4LibraryIcons";

export type IconEntity = IconEntity;
export type IconAttributes = IconAttributes;

export const iconDefaults = (): IconAttributes =>
  clone({
    name: "",
    texture: "",
    height: 16,
    width: 16
  });

export class Icon extends X4Entity<IconEntity> {
  get name() {
    return this.xmlDef.Attributes.name;
  }
  set name(v: string) {
    if (typeof v !== "string") throw new TypeError(`${v} must be string`);
    if (typeof v === "string") this.xmlDef.Attributes.name = v;
  }
  get texture() {
    return this.xmlDef.Attributes.texture || iconDefaults().texture;
  }
  set texture(v: string | null | undefined) {
    if (typeof v !== "string") throw new TypeError(`${v} must be string`);
    if (typeof v === "string") this.xmlDef.Attributes.texture = v;
  }
  get height() {
    return this.xmlDef.Attributes.height || iconDefaults().height;
  }
  set height(v: number | null | undefined) {
    if (undefined == v) delete this.xmlDef.Attributes.height;
    if (typeof v !== "number") throw new TypeError(`${v} must be number`);
    if (typeof v === "number") this.xmlDef.Attributes.height = v;
  }
  get width() {
    return this.xmlDef.Attributes.width || iconDefaults().width;
  }
  set width(v: number | null | undefined) {
    if (undefined == v) delete this.xmlDef.Attributes.width;
    if (typeof v !== "number") throw new TypeError(`${v} must be number`);
    if (typeof v === "number") this.xmlDef.Attributes.width = v;
  }
  get active() {
    return this.xmlDef.Attributes.active || iconDefaults().active;
  }
  set active(v: string | null | undefined) {
    if (undefined == v) delete this.xmlDef.Attributes.active;
    if (typeof v !== "string") throw new TypeError(`${v} must be string`);
    if (typeof v === "string") this.xmlDef.Attributes.active = v;
  }
  get inactive() {
    return this.xmlDef.Attributes.inactive || iconDefaults().inactive;
  }
  set inactive(v: string | null | undefined) {
    if (undefined == v) delete this.xmlDef.Attributes.inactive;
    if (typeof v !== "string") throw new TypeError(`${v} must be string`);
    if (typeof v === "string") this.xmlDef.Attributes.inactive = v;
  }
  get selected() {
    return this.xmlDef.Attributes.selected || iconDefaults().selected;
  }
  set selected(v: string | null | undefined) {
    if (undefined == v) delete this.xmlDef.Attributes.selected;
    if (typeof v === "string") this.xmlDef.Attributes.selected = v;
  }
  constructor(
    constructOptions: IconAttributes = iconDefaults(),
    xmlDef?: IconEntity
  ) {
    super(
      X4EntityType.LIBRARY_ICON,
      "icon",
      iconDefaults(),
      xmlDef || {
        Attributes: constructOptions
      }
    );
  }
}

export default Icon;
