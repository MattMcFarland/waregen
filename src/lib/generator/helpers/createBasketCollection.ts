import { GeneratorConfig } from "..";
import { Basket, BasketCollection } from "../../entities";
import { IdRoster } from "../../utils/IdRoster";

interface BasketHash {
  [key: string]: string[];
}

export default function createBasketCollection(
  config: GeneratorConfig
): BasketCollection {
  const baskets = new BasketCollection();
  const basketHash = <BasketHash>{};
  config.addwaresList.forEach(addware => {
    const ids = new IdRoster(addware.Attributes.id, config.modPrefix);
    const insertionList = addware.Attributes.baskets
      .replace("[", "")
      .replace("]", "")
      .split(",")
      .map(str => str.trim());
    insertionList.forEach(basketName => {
      if (!basketHash[basketName]) {
        basketHash[basketName] = [ids.ware];
      } else {
        basketHash[basketName].push(ids.ware);
      }
    });
  });
  Object.entries(basketHash).forEach(([id, wares]) => {
    baskets.add(new Basket({ id, wares }));
  });
  return baskets;
}
