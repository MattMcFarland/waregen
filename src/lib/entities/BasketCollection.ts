import { X4Entity, X4EntityType, XMLPatchTypes } from "./X4Entity";
import {
  BasketEntity,
  Baskets,
  WareEntity as BasketWareEntity,
  WaresEntityOrWareEntity as BasketEntityWares
} from "..//XMLTypes/X4LibraryBaskets";
import { BuilderOptions } from "../utils/xml";
import { Basket, BasketInterface } from "./Basket";

export type BasketWareEntity = BasketWareEntity;
export type BasketWareEntities = BasketWareEntity[];
export interface BasketIndex {
  [key: string]: Basket;
}
export class BasketCollection extends X4Entity<Baskets> {
  private basketIndex: BasketIndex = {};
  protected onImport(_xmlDef: Baskets) {
    delete this.basketIndex;
    this.basketIndex = {};
    delete this.xmlDef;
    this.xmlDef = { basket: [] };

    if (Array.isArray(_xmlDef.basket)) {
      _xmlDef.basket.forEach((basketEntity: BasketEntity) => {
        const constructorOptions: BasketInterface = {
          id: basketEntity.Attributes && basketEntity.Attributes.id,
          name:
            (basketEntity.Attributes && basketEntity.Attributes.name) ||
            undefined,
          description:
            (basketEntity.Attributes && basketEntity.Attributes.description) ||
            undefined,
          wares:
            (basketEntity.wares && importBasketWares(basketEntity.wares)) ||
            undefined
        };
        const basket = new Basket(constructorOptions);
        this.add(basket);
      });
    }
  }
  add = (basket: Basket) => {
    if (this.basketIndex[basket.id]) {
      throw new Error(
        `basket ${basket.id} already exists, did you mean replaceWare?`
      );
    }
    this.basketIndex[basket.id] = basket;
    if (Array.isArray(this.xmlDef.basket)) {
      this.xmlDef.basket = [...this.xmlDef.basket, basket.__xmlDef];
    }
  };
  remove = (id: string) => {
    if (!this.basketIndex[id]) {
      throw new Error(`cant remove basket ${id} because it does not exist`);
    }
    delete this.basketIndex[id];
    if (Array.isArray(this.xmlDef.basket)) {
      this.xmlDef.basket = this.xmlDef.basket.filter(
        basketEntity => basketEntity.Attributes.id !== id
      );
    }
  };
  public toXmlPatch(
    patchType: XMLPatchTypes,
    selector?: string,
    pretty: boolean = true,
    options: BuilderOptions = {}
  ) {
    options.rootName = "diff";
    const obj = {
      [patchType]: Object.values(this.basketIndex).map(basket => {
        return {
          ...basket.__xmlDef,
          Attributes: { sel: basket.selector }
        };
      })
    };
    return this.xmlBuild(obj, pretty, options);
  }
  replace = (basket: Basket) => {
    this.remove(basket.id);
    this.add(basket);
  };
  extract = (wareId: string) => {
    return this.basketIndex[wareId];
  };
  constructor(optionalBasketList?: Basket[]) {
    super(X4EntityType.LIBRARY_BASKETS, "baskets", {
      basket: []
    });
    if (Array.isArray(optionalBasketList)) {
      optionalBasketList.forEach((basket: Basket) => {
        this.add(basket);
      });
    }
  }
}

export function importBasketWares(wares: BasketEntityWares[]) {
  if (wares[0].ware) {
    return wares[0].ware.map(ware => ware.Attributes.ware);
  }
}

export default BasketCollection;
