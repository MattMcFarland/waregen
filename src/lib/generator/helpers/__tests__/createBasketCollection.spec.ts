import { mockGeneratorConfig } from "./util";
import createBasketCollection from "../createBasketCollection";
import { XMLPatchTypes } from "@@/entities/X4Entity";
describe("generator/helpers", () => {
  describe("createBasketCollection", () => {
    const mockConfig = mockGeneratorConfig();
    const basketCollection = createBasketCollection(mockConfig);
    test("toJson()", () => {
      expect(basketCollection.toJson()).toMatchSnapshot();
    });
    test("toXml()", () => {
      expect(basketCollection.toXml()).toMatchSnapshot();
    });
    test("toXmlPatch()", () => {
      expect(basketCollection.toXmlPatch(XMLPatchTypes.ADD)).toMatchSnapshot();
    });
  });
});
