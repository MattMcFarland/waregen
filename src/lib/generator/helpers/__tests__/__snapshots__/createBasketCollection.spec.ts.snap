// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator/helpers createBasketCollection toJson() 1`] = `"{\\"basket\\":[{\\"Attributes\\":{\\"id\\":\\"test_basket_a\\"},\\"wares\\":[{\\"ware\\":[{\\"Attributes\\":{\\"ware\\":\\"pfx_a\\"}}]}]},{\\"Attributes\\":{\\"id\\":\\"pirate_container\\"},\\"wares\\":[{\\"ware\\":[{\\"Attributes\\":{\\"ware\\":\\"pfx_a\\"}},{\\"Attributes\\":{\\"ware\\":\\"pfx_b\\"}},{\\"Attributes\\":{\\"ware\\":\\"pfx_c\\"}},{\\"Attributes\\":{\\"ware\\":\\"pfx_d\\"}},{\\"Attributes\\":{\\"ware\\":\\"pfx_e\\"}}]}]},{\\"Attributes\\":{\\"id\\":\\"all_container\\"},\\"wares\\":[{\\"ware\\":[{\\"Attributes\\":{\\"ware\\":\\"pfx_a\\"}},{\\"Attributes\\":{\\"ware\\":\\"pfx_b\\"}},{\\"Attributes\\":{\\"ware\\":\\"pfx_c\\"}},{\\"Attributes\\":{\\"ware\\":\\"pfx_d\\"}},{\\"Attributes\\":{\\"ware\\":\\"pfx_e\\"}}]}]},{\\"Attributes\\":{\\"id\\":\\"all\\"},\\"wares\\":[{\\"ware\\":[{\\"Attributes\\":{\\"ware\\":\\"pfx_a\\"}},{\\"Attributes\\":{\\"ware\\":\\"pfx_b\\"}},{\\"Attributes\\":{\\"ware\\":\\"pfx_c\\"}},{\\"Attributes\\":{\\"ware\\":\\"pfx_d\\"}},{\\"Attributes\\":{\\"ware\\":\\"pfx_e\\"}}]}]},{\\"Attributes\\":{\\"id\\":\\"test_basket_b\\"},\\"wares\\":[{\\"ware\\":[{\\"Attributes\\":{\\"ware\\":\\"pfx_b\\"}}]}]},{\\"Attributes\\":{\\"id\\":\\"test_basket_c\\"},\\"wares\\":[{\\"ware\\":[{\\"Attributes\\":{\\"ware\\":\\"pfx_c\\"}}]}]},{\\"Attributes\\":{\\"id\\":\\"test_basket_d\\"},\\"wares\\":[{\\"ware\\":[{\\"Attributes\\":{\\"ware\\":\\"pfx_d\\"}}]}]},{\\"Attributes\\":{\\"id\\":\\"test_basket_e\\"},\\"wares\\":[{\\"ware\\":[{\\"Attributes\\":{\\"ware\\":\\"pfx_e\\"}}]}]}]}"`;

exports[`generator/helpers createBasketCollection toXml() 1`] = `
"<baskets>
  <basket id=\\"test_basket_a\\">
    <wares>
      <ware ware=\\"pfx_a\\"/>
    </wares>
  </basket>
  <basket id=\\"pirate_container\\">
    <wares>
      <ware ware=\\"pfx_a\\"/>
      <ware ware=\\"pfx_b\\"/>
      <ware ware=\\"pfx_c\\"/>
      <ware ware=\\"pfx_d\\"/>
      <ware ware=\\"pfx_e\\"/>
    </wares>
  </basket>
  <basket id=\\"all_container\\">
    <wares>
      <ware ware=\\"pfx_a\\"/>
      <ware ware=\\"pfx_b\\"/>
      <ware ware=\\"pfx_c\\"/>
      <ware ware=\\"pfx_d\\"/>
      <ware ware=\\"pfx_e\\"/>
    </wares>
  </basket>
  <basket id=\\"all\\">
    <wares>
      <ware ware=\\"pfx_a\\"/>
      <ware ware=\\"pfx_b\\"/>
      <ware ware=\\"pfx_c\\"/>
      <ware ware=\\"pfx_d\\"/>
      <ware ware=\\"pfx_e\\"/>
    </wares>
  </basket>
  <basket id=\\"test_basket_b\\">
    <wares>
      <ware ware=\\"pfx_b\\"/>
    </wares>
  </basket>
  <basket id=\\"test_basket_c\\">
    <wares>
      <ware ware=\\"pfx_c\\"/>
    </wares>
  </basket>
  <basket id=\\"test_basket_d\\">
    <wares>
      <ware ware=\\"pfx_d\\"/>
    </wares>
  </basket>
  <basket id=\\"test_basket_e\\">
    <wares>
      <ware ware=\\"pfx_e\\"/>
    </wares>
  </basket>
</baskets>"
`;

exports[`generator/helpers createBasketCollection toXmlPatch() 1`] = `
"<diff>
  <add sel=\\"*/basket[@id='test_basket_a']\\">
    <wares>
      <ware ware=\\"pfx_a\\"/>
    </wares>
  </add>
  <add sel=\\"*/basket[@id='pirate_container']\\">
    <wares>
      <ware ware=\\"pfx_a\\"/>
      <ware ware=\\"pfx_b\\"/>
      <ware ware=\\"pfx_c\\"/>
      <ware ware=\\"pfx_d\\"/>
      <ware ware=\\"pfx_e\\"/>
    </wares>
  </add>
  <add sel=\\"*/basket[@id='all_container']\\">
    <wares>
      <ware ware=\\"pfx_a\\"/>
      <ware ware=\\"pfx_b\\"/>
      <ware ware=\\"pfx_c\\"/>
      <ware ware=\\"pfx_d\\"/>
      <ware ware=\\"pfx_e\\"/>
    </wares>
  </add>
  <add sel=\\"*/basket[@id='all']\\">
    <wares>
      <ware ware=\\"pfx_a\\"/>
      <ware ware=\\"pfx_b\\"/>
      <ware ware=\\"pfx_c\\"/>
      <ware ware=\\"pfx_d\\"/>
      <ware ware=\\"pfx_e\\"/>
    </wares>
  </add>
  <add sel=\\"*/basket[@id='test_basket_b']\\">
    <wares>
      <ware ware=\\"pfx_b\\"/>
    </wares>
  </add>
  <add sel=\\"*/basket[@id='test_basket_c']\\">
    <wares>
      <ware ware=\\"pfx_c\\"/>
    </wares>
  </add>
  <add sel=\\"*/basket[@id='test_basket_d']\\">
    <wares>
      <ware ware=\\"pfx_d\\"/>
    </wares>
  </add>
  <add sel=\\"*/basket[@id='test_basket_e']\\">
    <wares>
      <ware ware=\\"pfx_e\\"/>
    </wares>
  </add>
</diff>"
`;
